content/posts/2025-01-03-packets-xlation-enrichment.md

```diff
@@ -0,0 +1,197 @@
+---
+layout: :theme/post
+title: "Enhancing NetObserv for Kubernetes Service Flows using eBPF"
+description: NetObserv Packet Translation provide enriched service endpoint POD information
+tags: eBPF,Services,Monitoring,Troubleshooting
+authors: [msherif1234]
+---
+
+# Enhancing NetObserv for Kubernetes Service Flows using eBPF: Adding Translated Endpoint Information
+
+## Introduction
+
+In a cloud-native environment, Kubernetes has become the de facto standard for managing containerized workloads.
+While its service discovery and load-balancing features are robust, gaining visibility into the actual endpoints
+serving traffic often requires complex instrumentation or external tooling.
+This is where eBPF (extended Berkeley Packet Filter) shines, offering a way to deeply observe and enrich Kubernetes
+service flows without intrusive changes to your applications.
+
+In this blog, we'll explore how eBPF can be used to add translated endpoint (xlated endpoint) information to
+Kubernetes service flows, providing insights into backend behavior and improving observability.
+
+## The Challenge
+
+When traffic flows through a Kubernetes Service, it often gets abstracted away by kube-proxy or other load balancers.
+For example, a request to `my-service` on `ClusterIP` is transparently routed to one of the available pods.
+However, network observability will capture both the service traffic and the virtual service endpoint as two separate flows.
+This lack of granularity makes troubleshooting and optimization more challenging.
+
+To solve this, we can:
+
+Capture the network flows at the kernel level.
+
+Enrich them with translated endpoint information, showing not only the service but also the specific backend pod.
+
+## How eBPF Can Help
+
+You can execute custom programs using eBPF in the Linux kernel, making it an ideal tool for network observability.
+
+Notable benefits of using eBPF include the following:
+
+- Granular Observability: Directly see which pod served a request.
+
+- Low Overhead: Operates in the kernel with minimal impact on performance.
+
+- Dynamic Updates: Respond to changes in Kubernetes without requiring application restarts.
+
+- Simplified Architecture: No need for sidecars or intrusive network plugins.
+
+Here’s a high-level approach:
+
+1- attach eBPF Programs:
+Use eBPF programs to hook into kernel networking events, such as kprobe on functions like `nf_nat_manip_pkt`.
+which will enable network observability eBPF agent to learn about all network translations events
+done via [Linux Conntrack Tool](https://conntrack-tools.netfilter.org/manual.html)
+
+2- enrich Flow Logs:
+As network packets are processed, the eBPF hook will augment flow logs with metadata about the translated endpoint.
+
+This will include:
+
+- Source Pod IP
+
+- Source Port
+
+- Destination Pod IP
+
+- Destination Port
+
+- Conntrack Zone ID
```

Comment:
what is zone id?

Reply:
https://lwn.net/Articles/370152/#:~:text=A%20zone%20is%20simply%20a,to%20seperate%20conntrack%20defragmentation%20queues.